### DISCUSSION TITLE
ESP IDF v.5.2.

### DISCUSSION THREAD

**Discussion by muexxl** [UncleRus/esp-idf-lib#612] [Q&A] [answered]
Recently v5.2. of ESP IDF has been released. (https://github.com/espressif/esp-idf/releases/tag/v5.2 )

Therein, the I2C driver has been redesigned. (https://docs.espressif.com/projects/esp-idf/en/stable/esp32/migration-guides/release-5.x/5.2/peripherals.html) 

Does this affect esp-idf-lib?
Is it safe to migrate from 5.1.2. to 5.2.0 ?

Does anyone have already experience with the migration?

**Comment by Florida-Rob** [comment]
I also would like the answer to this.  I have seen many incompatibilities, and PANICS.  hence i think the answer is NO
the "i2c_cmd_handle_t" is deprecated since esp-idf 4.3, so what is the new command ? has anyone ported this library for esp-idf 5.2 or newer ?

**Comment by UncleRus** [comment]
Right now I have several devices working, the firmware for which was compiled two weeks ago using the latest ESP-IDF master at that time.
The devices use about ten I2C device drivers from the library. I did not find any deviations or bugs, except for the message
```
W (545) i2c: This driver is an old driver, please migrate your application code to adapt `driver/i2c_master.h`
```

**Comment by Florida-Rob** [comment]
[from the esp-idf:](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/i2c.html)
"Note: **The legacy driver can't coexist with the new driver**. Include i2c.h to use the legacy driver or the other two headers to use the new driver. **Please keep in mind that the legacy driver is now deprecated and will be removed in future**."
we need help here.  i have a lot of time invested in this driver here, and too many PANIC and incompatibilities.  will this driver be updated ? otherwise, i'm afraid its all dead...

**Reply by UncleRus** [reply]
> and too many PANIC and incompatibilities

I don't have any

**Reply by Florida-Rob** [reply]
It does with esp-idf 5.2+, I see I2C errors in LOGE (in red) very occasionally, which cause a panic and reboot. happened 5 times in tests over 2 weeks.  so i pulled the use of the driver, and they stopped (just PCA9557).  I have been without i2C as i worked on other code for 4 months, but now back re-adding I2C with the new esp-idf library, hence this request for a new driver.

**Reply by UncleRus** [reply]
In this case, please kindly confirm your statements: write minimal code that reproduces the errors, create an issue, show the logs, double-check your hardware (90% of errors are problems with the hardware). Until then, this will all remain strange alarmist statements. What, for example, does PANIC and reboot mean when using functions from i2clib? Are you wrapping calls to the driver in the ESP_ERROR_CHECK macro?

In general, I see two constructive options for continuing this discussion:

1. You begin work on porting i2clib, as I wrote earlier
2. You are creating a specific issue on a problem with the esp-idf-lib component. And keep in mind, before creating this issue, it is worth checking whether the error is really related to the ESP-IDF version, or whether your error is related to the hardware, or whether it is a bug of a specific driver.

**Comment by UncleRus** [comment, marked as answer]
I certainly agree that it’s time to port i2clib to a new driver (and at the same time maintain compatibility with older versions of ESP-IDF and ESP8266 RTOS SDK).

I invite the discussion participants (and everyone else) to initiate the process by creating a PR and starting work in it.

**Comment by Florida-Rob** [comment]
looks like the answers [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/i2c.html)
functions needed with use of "driver/i2c_master.h", are:
i2c_master_transmit_receive()
i2c_master_transmit()
i2c_master_receive()

**Reply by UncleRus** [reply]
No need to rub my nose into the Espressif documentation, I can read it perfectly well without such condescending links.
Once again: either you start porting work yourself, or politely ask those who can do it for you.

**Comment by Florida-Rob** [comment]
i ported the code for the esp-idf 5.2, replaces deprecated i2c.h library for new i2c_master.h
its not a direct replacement, but close, how would you like me to share the file ?

**Reply by ryanminnig** [reply]
Is this something you would be willing to share?  I started doing a port and quickly got discouraged by the number of errors during compilation.

**Reply by Florida-Rob** [reply]
sure thing.  i ported the core, then libraries for the SSD1306 display, and PCA9557 - but anything really would port fine with the new core.
Not sure how to share the files at this point, but attached
[new_i2c.zip](https://github.com/user-attachments/files/16952391/new_i2c.zip)

#include "new_i2c.h"
#include "ssd1306.h"
#include "pca9557.h"
